This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests
import collections
import json

Note: API's can change a bit with each version, for this exercise it is reccomended to use the "V3" quandl api at `https://www.quandl.com/api/v3/`

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?api_key='+API_KEY

In [4]:
res = requests.get(url)

In [5]:
print(res)

<Response [200]>


In [56]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json = res.json()


In [8]:
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=' + API_KEY

In [9]:
res = requests.get(url)

In [10]:
print(res)

<Response [200]>


In [57]:
json = res.json()


In [17]:
print(type(json))

<class 'dict'>


In [13]:
#Task 3
list_of_data = json['dataset']['data']

In [14]:
def make_list(index):
    new_list = [list[index] for list in list_of_data]
    new_list = np.array(new_list, dtype=np.float)
    return new_list

In [15]:
print(json['dataset']['column_names'])

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


In [18]:
import numpy as np

In [19]:
opening_prices = make_list(1)

In [20]:
print(len(opening_prices))

255


In [21]:
print(np.nanmax(opening_prices))

53.11


In [22]:
print(np.nanmin(opening_prices))

34.0


# Task 3 Answer
Highest opening price was 53.11<br>Lowest opening price was 34.00

In [23]:
high_prices = make_list(2)

In [24]:
low_prices = make_list(3)

In [25]:
print(len(high_prices))

255


In [26]:
print(len(low_prices))

255


In [27]:
diffs = high_prices - low_prices

In [28]:
print(len(diffs))

255


In [29]:
print(high_prices[:10])

[51.94 51.82 51.89 51.5  51.52 52.04 52.73 52.75 52.7  53.54]


In [30]:
print(low_prices[:10])

[51.45 51.43 50.76 50.92 50.9  51.2  51.07 51.61 51.64 52.15]


In [31]:
print(diffs[:10])

[0.49 0.39 1.13 0.58 0.62 0.84 1.66 1.14 1.06 1.39]


In [32]:
print(np.nanmax(diffs))

2.8100000000000023


# Task 4 Answer
Largest change was 2.2.8100000000000023

In [37]:
closing_prices = make_list(4)

In [38]:
print(len(closing_prices))

255


In [39]:
print(closing_prices[:10])

[51.76 51.6  51.82 51.32 51.4  51.27 51.66 52.62 52.01 52.67]


In [40]:
daily_changes = np.diff(closing_prices)

In [41]:
print(len(daily_changes))

254


In [42]:
print(daily_changes[:10])

[-0.16  0.22 -0.5   0.08 -0.13  0.39  0.96 -0.61  0.66  0.42]


In [43]:
print(np.nanmax(daily_changes))

2.559999999999995


# Task 5 Answer
Largest change was 2.559999999999995

In [44]:
volumes = make_list(6)

In [45]:
print(len(volumes))

255


In [46]:
sum_vols = np.sum(volumes)

In [47]:
print(np.isnan(sum_vols))

False


In [49]:
print(np.mean(volumes))

89124.33725490196


In [50]:
print(volumes[:10])

[ 34640.  40660.  57452.  71165. 120649.  50587. 137313. 129733. 204080.
 132981.]


# Task 6 Answer
Average daily volume is 89124.33725490196

In [52]:
median_volume = np.median(np.sort(volumes))

In [53]:
print(median_volume)

76286.0


In [55]:
print(np.sort(volumes)[127])

76286.0


# Task 7 Answer
Median trading volume is 76286.0

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)